### ISSUE TITLE
Add support for multiple sequence alignment, homology modeling, and deep structural prediction 

### ISSUE THREAD

**Issue by rbharath** [deepchem/deepchem#2150]
Introduction
------------

DeepChem's support for protein-based deep learning is still pretty early-stage. In this issue I propose a framework for adding support to DeepChem for doing more systematic machine learning with proteins. In particular, I propose the addition of new featurizers which compute multiple sequence alignment, mutual-information/multiple contact prediction,  features. I also propose the addition of a new homology modeling class which allows the construction of homology models with a DeepChem API. 

These proposed modifications are inspired by @miaecle's work on https://github.com/miaecle/PNet. By bringing in some of PNet's infrastructure into DeepChem, we can leverage DeepChem's infrastructure to build larger scale deep models for protein structure prediction. We should also be able to build protein structure predictions models in DeepChem as is possible in PNet (https://github.com/miaecle/PNet/blob/master/examples/atrous_conv_example.py)

Featurizers
-----------

I propose the addition/modification of four featurization classes to DeepChem:

- `dc.feat.MultipleSequenceAlignmentFeaturizer`: This featurizer would compute multiple sequence alignments for a given sequence. The implementation should closely follow that of PNet https://github.com/miaecle/PNet/blob/master/pnet/feat/MSA.py#L30. Note that this depends on HHLBits under the hodd.
- `dc.feat.OneHotFeaturizer`: Generalize this class so that it works with protein sequences and not only small molecules. (We might as well generalize this to work with arbitrary sequences while we're at it)
- `dc.feat.MutuaInformationFeaturizer`: This featurizer would compute mutual information for a given sequence. The implementation should closely follow that of PNet https://github.com/miaecle/PNet/blob/master/pnet/feat/twoD_features.py#L43.
- `dc.feat.MeanContactPotentialFeaturizer`: This featurizer would compute mean contact potential for a given sequence. The implementation should closely follow that of PNet https://github.com/miaecle/PNet/blob/master/pnet/feat/twoD_features.py#L43.

MoleculeNet Datasets
----------------------

I propose that we add 4 new datasets to MoleculeNet following the PNet implementation:

- CATH: https://github.com/miaecle/PNet/blob/master/pnet/utils/sequence_utils.py#L73
- PDB50: https://github.com/miaecle/PNet/blob/master/pnet/utils/sequence_utils.py#L80
- CAMEO: https://github.com/miaecle/PNet/blob/master/pnet/utils/sequence_utils.py#L105
- CASP: (With sub datasets CASP5-CASP12) https://github.com/miaecle/PNet/blob/master/pnet/utils/sequence_utils.py#L55

Metrics
--------

I propose we add the following metric:

- `dc.metrics.top_k_accuracy`: https://github.com/miaecle/PNet/blob/master/pnet/utils/metrics.py#L176

Homology Modelling
---------------------

At present the most popular tool for homology modelling is the [Modeller package](https://salilab.org/modeller/). This package is well tested and widely used, but has the difficult limitation that it is only free for academic non-profit users and isn't open source. Given this limitation, I propose that it would be useful to create the capability to generate homology models in DeepChem. For this purpose I propose the addition of a new class to DeepChem

- `dc.dock.HomologyModeller`: Underneath the hood, this class will use `hhmakemodel.py` from `hhsuite` and `MultipleSequenceAlignmentFeaturizer` to construct homology models for a given sequence.

Models
-------

I'm not yet sure what models we should add pre-built, but here is an example of PNet's models that could serve as inspiration: https://github.com/miaecle/PNet/blob/master/pnet/models/conv_net_contact_map.py

License Considerations
-----------------------
An important consideration here is that `hh-suite` is GPL licensed. After some good discussion with @peastman and looking through the GPL FAQ's, I believe that we can create MIT licensed code in DeepChem that links to the GPL hhsuite. See these FAQ pointers from @peastman:

- https://www.gnu.org/licenses/gpl-faq.html#WhatIsCompatible
- https://www.gnu.org/licenses/gpl-faq.html#WhatDoesCompatMean
- https://www.gnu.org/licenses/gpl-faq.html#LinkingWithGPL

Any particular DeepChem installation which also has hh-suite installed would become a GPL program, but the DeepChem source code itself could remain MIT licensed since MIT license is GPL compatible

Development Plan
------------------
For now this is a preliminary design I'm posting for feedback and comments from folks. @miaecle @peastman I'd love your comments on this design in particular! If this makes sense, we can then plan on the development roadmap for getting these features into DeepChem itself.

**Comment by rbharath** [comment]
Adding on a couple useful links after offline discussion with @miaecle:

1. https://github.com/deepmind/deepmind-research/tree/master/alphafold_casp13: DeepMind's code release. It's incomplete since it doesn't have the full pipeline for AlphaFold
2. https://github.com/Urinx/alphafold_pytorch: A more complete attempt to reconstruct the AlphaFold pipeline

**Comment by miaecle** [comment]
@rbharath Thanks for raising this thread! Regarding the contact map (distance map) prediction step, here are some basic framework for featurizer setup:

Input features (1D)
--------------------

1. Amino acid type (`dc.feat.OneHotFeaturizer`)
2. **Multiple sequence alignment** (`dc.feat.MultipleSequenceAlignmentFeaturizer`)
    + Sequence profile
    + Deletion prob
    + Num alignments
3. Secondary structure (optional)
4. Solvent accessibility (optional)

Input features (2D)
--------------------

1. Mutual information

2. Mean contact potential

3. **Pseudo-likelihood** (this could be either coded in python or reusing external software such as CCMpred, plmDCA)

Labels
-------

1. alpha-carbon coordinates
2. beta-carbon coordinates
3. phi angle
4. psi angle
5. Pairwise distance matrix


I will work on porting the existing ones to current deepchem framework. @rbharath Could you point me one example featurizer class in deepchem that I could use for reference?

**Comment by rbharath** [comment]
@miaecle Really excited you can help out! I'd recommend using the `CircularFingerprint` featurizer as a template https://github.com/deepchem/deepchem/blob/master/deepchem/feat/molecule_featurizers/circular_fingerprint.py#L12.

One thing to consider is whether we want to add a new abstract superclass. `CircularFingerprint` inherits from `MolecularFeaturizer` which assumes its input is a molecule represented by a smiles string. We might want to consider adding an abstract `ProteinFeaturizer` superclass more suited for featurizing proteins. CC @Mickdub

**Comment by miaecle** [comment]
@rbharath Yes I think it is better to have a separate super class. A related question is that do we have a standard dataset class for iteration in model training? (I could also port the dataset class I used in my repo as well)

**Comment by rbharath** [comment]
@miaecle Yep, check out the `Dataset` classes (https://deepchem.readthedocs.io/en/latest/datasets.html). I think these should be able to do what the PNet dataset class does (and are nicely optimized with things like a cache on `DiskDataset` for efficiency)

**Comment by peastman** [comment]
The [new version of AlphaFold](https://deepmind.com/blog/article/alphafold-a-solution-to-a-50-year-old-grand-challenge-in-biology) was just announced at CASP14.  It looks like a huge leap over their previous version.  They don't have a paper yet, but it should come eventually.  The abstract provides just enough information to be intriguing.

> We found that existing deep-learning architectures overly favor sequence-local interactions and  do  not  sufficiently  account  for  global  structural  constraints.  To  remedy  this,  we  have  developed a novel, attention-based deep learning architecture to achieve self-consistent structure prediction.  We  also  allow  the  deep  learning  algorithm  to  attend  arbitrarily  over  the  full  MSA  instead  of  using  pairwise  co-evolution  features  like  mutual  information  or  pseudolikelihood,  allowing the algorithm to ignore irrelevant sequences as well as to extract much richer information from  the  MSA.  The  resulting  algorithm  shows  vastly  improved  performance,  especially  for  shallow MSA depths, when compared to traditional co-evolution methods.

**Comment by peastman** [comment]
The AlphaFold 2 code has finally been released!  With the Apache license, so it's very permissive.  https://github.com/deepmind/alphafold

**Comment by rbharath** [comment]
That paper is really cool! It might be good to incorporate more infrastructure to predict RNA structures into DeepChem
